sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CombineJoinedAggregates.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, And, Attribute, AttributeMap, Expression, NamedExpression, Or}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.catalyst.plans.{Cross, FullOuter, Inner, JoinType, LeftOuter, RightOuter}
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, Filter, Join, LeafNode, LogicalPlan, Project, SerializeFromObject}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern.{AGGREGATE, JOIN}
+
+/**
+ * This rule eliminates the [[Join]] if all the join side are [[Aggregate]]s by combine these
+ * [[Aggregate]]s. This rule also support the nested [[Join]], as long as all the join sides for
+ * every [[Join]] are [[Aggregate]]s.
+ *
+ * Note: this rule doesn't following cases:
+ * 1. One of the to be merged two [[Aggregate]]s with child [[Filter]] and the other one is not.
```

Comment:
Thinking more about it, I think what really matters is the predicate selectivity. No filter means an always true predicate.

To be safe, I think we should reuse some code of DPP that determines selective predicates, and only merge aggregates if they both have selective predicates.

Reply:
Good idea.

Reply:
> Thinking more about it, I think what really matters is the predicate selectivity. No filter means an always true predicate.
> 
> To be safe, I think we should reuse some code of DPP that determines selective predicates, and only merge aggregates if they both have selective predicates.

Is that so? I think the performance gain of merging queries comes from the fact that we don't need to scan the same data 2 times but we can calculate the 2 aggregates in 1 run.
E.g. If we have 2 queries:
```
SELECT sum(a) FROM t WHERE <condition 1>
```
and
```
SELECT sum(b) FROM t WHERE <condition 2>
```
and we want to decide when merging the queries into
```
SELECT
  sum(a) FILTER (WHERE <condition 1>),
  sum(b) FILTER (WHERE <condition 2>)
FROM t
WHERE <condition 1> OR <condition 2>
```
make sense then I would say:
- if the conditions are true literals (or they doesn't exist at all)
- or the conditions match

then it definitely does make sense to merge them.
But in the first case the conditions are not selective at all and in the second case condition selectivity doesn't matter. (Actually these are the 2 cases that are already covered by `MergeScalarSubqueries.tryMergePlans()`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala#L275-L286.)

Now, if the conditions differ then it isn't trivial to calculate if scans overlap, but I think if both condisions are highly selective then we have less chance for scans to overlap, don't we?

In my https://github.com/apache/spark/pull/37630 I used a different heuristics to disable merging of aggregates with different filter conditions. If the conditions contain any partitioning or bucketing columns then aggregates are not merged.

Reply:
Actually, if we disable this feature by default then do we need to define any heuristics? Can't we just let the user enable the feature if needed?

Reply:
> make sense then I would say:
>

> - if the conditions are true literals (or they doesn't exist at all)

> - or the conditions match

I agree the opinion. This PR support these two cases.

1. The predicate of two side are true or they doesn't exist at all.
2. The predicate of two side are selective.

Reply:
Please see `test("join two side are Aggregates without Filter")`.

Reply:
Hm, I argued against 2., please see above.

Reply:
>  Now, if the conditions differ then it isn't trivial to calculate if scans overlap, but I think if both conditions are highly > > selective then we have less chance for scans to overlap, don't we?

This comment is the same 2 ?

Reply:
let's think about the benefits of merging aggregates
1. less scan traffic (overlapped scans are merged)
2. batch execution of aggregate functions (one Aggregate operator to run all aggregate functions)

That said, it's always beneficial to merge aggregates? I think the assumption is, if we scan a table N times with N different filters `F1`, `F2`, ... `Fn`, the scan traffic is the same or less if we scan the table once with `F1 OR F2 OR ... Fn`.

Reply:
> disabled merge (to be on the safe side) if any filters are pushed down to scans as partitioning or bucketing filters

The treatment looks too strict. For partition filters, as long as the number of data rows is not too large, it can still benefit.

Reply:
> allowed merge depending on a new spark.sql.planMerge.ignorePushedDataFilters

The treatment looks too strict too. Even if the data filter cannot be pushed down, we can still achieve many benefits.

Reply:
> > disabled merge (to be on the safe side) if any filters are pushed down to scans as partitioning or bucketing filters
> 
> The treatment looks too strict. For partition filters, as long as the number of data rows is not too large, it can still benefit.

Sure, merging can be beneficial (partitions filters overlap or scan cost is not significant), but I wanted to be on the safe side with the PR. We can also lift that restriction later with a more sophisticated logic.

> > allowed merge depending on a new spark.sql.planMerge.ignorePushedDataFilters
> 
> The treatment looks too strict too. Even if the data filter cannot be pushed down, we can still achieve many benefits.

If no filters are pushed down then merging is allowed, see 1st point.

Reply:
I've just updated the last sentence to make it clear.

Reply:
@peter-toth After the discussion offline between @cloud-fan and me, there is a way to limit the merged number.
So the merged filter won't be too bulky, and the amount of data repeatedly processed by the filter won't be too large.

Reply:
Do you mean the merged number of aggregates? How does that help to avoid https://github.com/apache/spark/pull/42223#discussion_r1282023520?

Reply:
I think a difference here is this PR only merges Aggregates with this pattern
```
Aggregate
  Filter
    TableScan
```

That said, `that data "flows through" the whole plan and gets filtered up in the aggregate node only` is not an issue as the table scan output goes directly to Aggregate and gets filtered.

Reply:
I think it's a bit hard to estimate the data size of "overlapped scans". The only safe case is when one filter is always `true`. For `p1 = 1` and `p2 = 1`, if we don't know the correlation of column `p1` and `p2`, we don't know the overlapped scan size.

Another direction is we stop merging aggregates if the filter is costly enough. We can define that a filter expression is cheap if it only contains simple comparison checks and the expression tree size is smaller than a threshold.

Reply:
> I think a difference here is this PR only merges Aggregates with this pattern
> 
> ```
> Aggregate
>   Filter
>     TableScan
> ```
> 
> That said, `that data "flows through" the whole plan and gets filtered up in the aggregate node only` is not an issue as the table scan output goes directly to Aggregate and gets filtered.

That's true, merging logic in this PR is simpler (e.g. can't handle joins). But please note that currently this PR accepts different patterns as well. Infinite number of `Projects` and `Filter` nodes are accepted between the `Aggregate` and `Scan` nodes as long as both queries contains them. Although that's a bit atificial example as optimizer rules gets rid of them in most cases...
(@beliefer, this is actually connected to https://github.com/apache/spark/pull/42223#discussion_r1279051263 that still doesn't seem correct.)
But I'm Ok with accepting only this simple pattern like aggregates for merging.

> I think it's a bit hard to estimate the data size of "overlapped scans". The only safe case is when one filter is always `true`. For `p1 = 1` and `p2 = 1`, if we don't know the correlation of column `p1` and `p2`, we don't know the overlapped scan size.

I can't agree more.

> Another direction is we stop merging aggregates if the filter is costly enough. We can define that a filter expression is cheap if it only contains simple comparison checks and the expression tree size is smaller than a threshold.

Hm, or we could also avoid double evaluation by aliasing conditions like:
```
Aggregate sum(a) FILTER (WHERE condition1), sum(b) FILTER (WHERE condition2)
  Filter condition1 OR condition2
    Project <condition 1> as condition1, <condition 2> as condition2
      TableScan
```

Reply:
Indeed, the suggested merged is worse than the original individual queries when scans don't overlap.

Reply:
> merged example in the ticket description is wrong.

I'm sorry. I wrote the example mistake. I corrected it already.

Reply:
```
Aggregate sum(a) FILTER (WHERE condition1), sum(b) FILTER (WHERE condition2)
  Filter condition1 OR condition2
    Project <condition 1> as condition1, <condition 2> as condition2
      TableScan
```

@peter-toth Put the filter condition into project list will inflate data and take shuffle more stressful.

Reply:
Merging can always "inflate" data in case of aggregate function columns are different to filter condition columns as those condition columns need to go up to aggregate in the merged query while in the original queries they were pruned after filter. In this case if we alias the conditions to small boolean columns then shuffle data size decrease and less "stressful".

The worst case you are referring to is when aggregate function columns are the superset of filter condition columns, but even in this case only small boolean columns are added and the gain is that we don't need to evaluate the filters twice.

Reply:
Take an example, condition 1 is `flag is true` and condition 2 is `flag is false`, put these conditions into project list will add two boolean columns.

Reply:
If a conditions is `f(c)` and `c` is not used in the aggregate function (e.g. it is `sum(a)`) then the project also removes `c` besides adding a new boolean column (we can write `sum(a) FILTER (WHERE <boolean column>)` instead of `sum(a) FILTER (WHERE f(c))`) .
If (`c1, c2, ..., c10)` is the condition and none of these are in the aggregate function then the project removes 10 columns and adds only 1 boolean.
What you are reffering to is the worst case, but IMO on average, the extra projects helps.

Reply:
In most cases, the `f(c)` are different. e.g. `f1(c)`, `f2(c)` and so on.

Reply:
That's ok, it is not the conditions (`f(c)`s) relation to each other what matters, but the filter columns set (`c1` ... `c10`) relation to aggregate column set (`a`).